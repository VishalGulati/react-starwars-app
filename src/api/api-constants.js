export const API_BASE_URL = 'https://swapi.co/api';
export const PLANETS_API = '/planets/';

// error codes
export const SERVER_ERROR_CODE = '000';
