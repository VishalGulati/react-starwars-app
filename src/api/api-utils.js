import { SERVER_ERROR_CODE } from './api-constants';

export const options = (method = 'GET', body = null) => {
  let data = {
    method,
    mode: 'cors',
    headers: {
      Accept: 'application/json',
      'Content-Type': 'application/json'
    }
  };

  if (body) {
    data.body = JSON.stringify(body);
  }

  return data;
};

export const apiCall = (url, apiOptions) =>
  fetch(url, apiOptions)
    .then(res => {
      if (res.status >= 200 && res.status < 300) {
        return res
          .json()
          .then(response => ({
            payload: response
          }))
          .catch(error => ({
            error
          }));
      } else {
        // Fail if status code is above 300
        return res.text().then(response => ({
          error: response
        }));
      }
    })
    .then(response => {
      if (response.payload && !response.error) {
        return response.payload
      } else {
        // if response contains error, that means we got error
        let errorResponse = {
          message: response.error,
          code: SERVER_ERROR_CODE
        };
        throw new Error(JSON.stringify(errorResponse));
      }
    })

    .catch(error => {
      // we don't want to throw JavaScript with simple message
      let errorResponse = {
        message: error.message,
        code: SERVER_ERROR_CODE
      };
      throw new Error(JSON.stringify(errorResponse));
    });
