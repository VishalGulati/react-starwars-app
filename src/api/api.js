import { API_BASE_URL, PLANETS_API } from './api-constants';
import { apiCall } from './api-utils';

// authenticate
export const authenticate = ({ username, password }) => {
    let auth = new Promise((resolve, reject) => {
        if (
            username.toLowerCase().indexOf('luke') >= 0 &&
            password.toLowerCase() === '19bby'
        )
            resolve({ status: 200, data: 'authenticated' });
        else
            reject({ status: 400, error: { code: 400, message: 'user not found!' } });
    });
    return auth;
};

// search api
export const getSearchResults = searchText =>
    apiCall(`${API_BASE_URL}${PLANETS_API}?search=${searchText}`);
