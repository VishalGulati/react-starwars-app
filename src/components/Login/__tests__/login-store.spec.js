import { put, call } from 'redux-saga/effects'

import {
    LOGIN_ACTION,
    LOGIN_SUCCESS_ACTION,
    LOGIN_FAIL_ACTION
} from '../login-constants';
import loginReducer, {
    loginAction,
    loginSuccess,
    loginSaga
} from '../login-store'
import { authenticate } from '../../api';

const initialState = {
    isLoading: false,
    data: null,
    error: null
}

const user = { username: 'luke', password: '19bby' };

describe('LoginStore', () => {
    it('should send login request properly', () => {
        const gen = loginSaga(loginAction(user));
        expect(gen.next().value).toEqual(call(authenticate, user))
        const data = { status: 200, data: 'authenticated' };
        expect(gen.next(data).value).toEqual(put(loginSuccess(data)))
        expect(gen.next().done).toBe(true)
    })
})
