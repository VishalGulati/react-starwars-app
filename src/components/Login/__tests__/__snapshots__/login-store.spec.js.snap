// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginStore send login action properly 1`] = `
Object {
  "data": null,
  "error": null,
  "isLoading": true,
}
`;
